pkg/ovs/ovsconfig/ovs_client.go

```diff
@@ -308,32 +309,62 @@ func (br *OVSBridge) CreateInternalPort(name string, ofPortRequest int32, extern
 	return br.createPort(name, name, "internal", ofPortRequest, externalIDs, nil)
 }
 
-// CreateVXLANPort creates a VXLAN tunnel port with the specified name on the
-// bridge.
+// CreateTunnelPort creates a tunnel port with the specified name and type on
+// the bridge.
 // If ofPortRequest is not zero, it will be passed to the OVS port creation.
-// If remoteIP is not empty, it will be set to the tunnel port interface
-// options; otherwise flow based tunneling will be configured.
-func (br *OVSBridge) CreateVXLANPort(name string, ofPortRequest int32, remoteIP string) (string, Error) {
-	return br.createTunnelPort(name, "vxlan", ofPortRequest, remoteIP)
+func (br *OVSBridge) CreateTunnelPort(name string, tunnelType TunnelType, ofPortRequest int32) (string, Error) {
+	return br.createTunnelPort(name, tunnelType, ofPortRequest, "", "", nil)
 }
 
-// CreateGenevePort creates a Geneve tunnel port with the specified name on the
-// bridge.
+// CreateTunnelPortExt creates a tunnel port with the specified name and type
+// on the bridge.
 // If ofPortRequest is not zero, it will be passed to the OVS port creation.
 // If remoteIP is not empty, it will be set to the tunnel port interface
 // options; otherwise flow based tunneling will be configured.
-func (br *OVSBridge) CreateGenevePort(name string, ofPortRequest int32, remoteIP string) (string, Error) {
-	return br.createTunnelPort(name, "geneve", ofPortRequest, remoteIP)
+// psk is for the pre-shared key of IPSec ESP tunnel. If it is not empty, it
+// will be set to the tunnel port interface options. Flow based IPSec tunnel is
+// not supported, so remoteIP must be provided too when psk is not empty.
+// If externalIDs is not nill, the IDs in it will be added to the port's
+// external_ids.
+func (br *OVSBridge) CreateTunnelPortExt(
+	name string,
+	tunnelType TunnelType,
+	ofPortRequest int32,
+	remoteIP string,
+	psk string,
+	externalIDs map[string]interface{}) (string, Error) {
+	if psk != "" && remoteIP != "" {
+		return "", newInvalidArgumentsError("IPSec tunnel can not be flow based. remoteIP must not be set")
+	}
+	return br.createTunnelPort(name, tunnelType, ofPortRequest, remoteIP, psk, externalIDs)
 }
 
-func (br *OVSBridge) createTunnelPort(name, ifType string, ofPortRequest int32, remoteIP string) (string, Error) {
-	var options map[string]interface{}
+func (br *OVSBridge) createTunnelPort(
+	name string,
+	tunnelType TunnelType,
+	ofPortRequest int32,
```

Comment:
do we need to validate this to be between 1 and 65,279? I'm sure the transaction will fail with error if an invalid ofport is given, but perhaps is "golang-ic" to validate beforehand?

In that case the type could be uint16 to safe a few bytes (unless the ovs tx expects a 32 bit integer)

Reply:
Sure. Will add the check.
OFPort in OVS schema is indeed int32. I would keep the same type (also in most cases I guess uint16 wont really save bytes).

Reply:
Are there some cases that the ofportRequest should be 0/-1, or I should say the caller expects OVS to assign ofport number automatically? One case I can think about is we might have multiple tunnel ports after enable IPSec, and let OVS to descides the ofport number should be better than Antrea client assign it.

Reply:
Yes. When ofPortRequest is 0, OVS will allocate the number. It is explained in the comments (maybe we could revise the comments to make it more clear though).